[Merged by Bors] - feat(analysis/complex/isometry): add linear_isometry_complex

frankymacster opened the pull request:
add proof about the isometries in the complex plane

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: first pass

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: rename assumptions, functions

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: move conj_one, conj_sub to data/complex/basic

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: add file header

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: add module docstrings

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: fix llinting errors

hrmacbeth commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,171 @@
+end
+
+lemma linear_isometry_complex (z : ℂ) (f : ℂ →ₗᵢ[ℝ] ℂ) :
+  ∃ a : ℂ, |a| = 1 ∧ ((f z = a * z) ∨ (f z = a * conj z)) :=
```
This statement is not as strong as it should be -- can you adjust it?  You've stated a theorem in which the `a` depends on the `z`, but in fact the `a` should be independent of `z`:
```lean
lemma linear_isometry_complex (f : ℂ →ₗᵢ[ℝ] ℂ) :
  ∃ a : ℂ, |a| = 1 ∧ ((∀ z, f z = a * z) ∨ (∀ z, f z = a * conj z)) :=
```

hrmacbeth replied:
Now I think of it, it might be useful in any case to give a separate definition of the rotation linear-isometries, and this allows for a more elegant statement (but equivalent to the other one I said):
```lean
def rotation {a : ℂ} (ha : |a| = 1) : ℂ →ₗᵢ[ℝ] ℂ :=
{ to_fun := λ z, a * z,
  etc }

lemma linear_isometry_complex (f : ℂ →ₗᵢ[ℝ] ℂ) :
  ∃ (a : ℂ) (ha : |a| = 1), f = rotation ha ∨ f = (rotation ha).comp conj_li :=
```

francoissunatori replied:
> This statement is not as strong as it should be -- can you adjust it? You've stated a theorem in which the `a` depends on the `z`, but in fact the `a` should be independent of `z`:
> 
> ```lean
> lemma linear_isometry_complex (f : ℂ →ₗᵢ[ℝ] ℂ) :
>   ∃ a : ℂ, |a| = 1 ∧ ((∀ z, f z = a * z) ∨ (∀ z, f z = a * conj z)) :=
> ```

I made the quoted change, however I haven't yet made [this change to using `rotation`](https://github.com/leanprover-community/mathlib/pull/6923#discussion_r602892917). Can I get what I have now reviewed before moving on to the next change? Or is it preferable to wait for the change to using `rotation`? Thanks!

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex:
 generalize to
  `∃ a : ℂ, |a| = 1 ∧ ((∀ z, f z = a * z) ∨ (∀ z, f z = a * conj z))`
(https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/linear_isometry_complex/near/234974756)

frankymacster committed:
Merge branch 'master' into linear_isometry_complex

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: fix llinting errors

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: add `rotation` definition

frankymacster committed:
Merge branch 'master' into linear_isometry_complex

frankymacster committed:
Merge branch 'master' into linear_isometry_complex

sgouezel commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,185 @@
+Copyright (c) 2021 François Sunatori. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ruben Van de Velde, Yakov Pechersky, Heather Macbeth, Kevin Buzzard, Greg Price,
+François Sunatori
```
You don't need to put everyone as an author here, putting your name is enough (it is not like, credits, but rather: who should we bug if something breaks in this file later on).

sgouezel commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,185 @@
+
+@[simp]
+lemma linear_isometry.id_to_linear_map : (linear_isometry.id.to_linear_map :
+  ℂ →ₗ[ℝ] ℂ) = linear_map.id := rfl
```
It is not obvious to me that this belongs to this file, and also that this is stated in the right generality. For instance, wouldn't this be true over any complex vector space instead of just `ℂ`?

sgouezel commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,185 @@
+lemma linear_isometry.id_to_linear_map : (linear_isometry.id.to_linear_map :
+  ℂ →ₗ[ℝ] ℂ) = linear_map.id := rfl
+
+lemma hf_re (f : ℂ →ₗᵢ[ℝ] ℂ) (h₃ : ∀ z, z + conj z = f z + conj (f z)) (z : ℂ) : (f z).re = z.re :=
```
The name of this lemma is not right: we don't know what `hf` is. Something like `re_apply_eq_re_of_add_conj_eq` would be better. More importantly, this is a rather specialized statement, so it should definitely not go in the root namespace. Why not put everything in the namespace `linear_isometry`?

The same applies to all the other lemmas in this file.

sgouezel commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,185 @@
+begin
+  simpa [ext_iff, add_re, add_im, conj_re, conj_im, ←two_mul,
+         (show (2 : ℝ) ≠ 0, by simp [two_ne_zero'])] using (h₃ z).symm
+end
```
If you just have one tactic, you can replace the `begin ... end` block with `by`, as in
```suggestion
by simpa [ext_iff, add_re, add_im, conj_re, conj_im, ←two_mul,
         (show (2 : ℝ) ≠ 0, by simp [two_ne_zero'])] using (h₃ z).symm
```

sgouezel commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,185 @@
+lemma linear_isometry.l0 {f : ℂ →ₗᵢ[ℝ] ℂ} (h : f 1 = 1) (z : ℂ) : ∥f z - 1∥ = ∥z - 1∥ := begin
+  rw ←linear_isometry.norm_map f (z - 1),
+  rw linear_isometry.map_sub,
+  rw h,
```
Successive rewrites can be compressed into one single rewrite, as in
```suggestion
  rw [←linear_isometry.norm_map f (z - 1), linear_isometry.map_sub, h]
```
(untested). And then you can replace the `begin ... end` block with a `by`.

The same comment applies several times later on in this file.

sgouezel commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,185 @@
+  rw [add_comm, ←this, add_comm],
+end
+
+lemma linear_isometry.re {f : ℂ →ₗᵢ[ℝ] ℂ} (h : f 1 = 1) (z : ℂ) : (f z).re = z.re := begin
```
Usually, the `begin` is alone on its line, not at the end of the previous line.

sgouezel commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,185 @@
+end
+
+lemma linear_isometry_complex_aux (f : ℂ →ₗᵢ[ℝ] ℂ) (h : f 1 = 1) :
+  (∀ z, f z = z) ∨ (∀ z, f z = conj z) := by {
```
A long proof like that is usually in a `begin ... end` block: `by` is only for very short proofs (one-liners or one single tactic invocation).
```suggestion
  (∀ z, f z = z) ∨ (∀ z, f z = conj z) :=
begin
```
(don't forget to adjust the `end` at the end of the proof). Same thing for the next lemma.

sgouezel commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,185 @@
+  let a := f 1,
+  use a,
+  split,
+    {
```
Braces are not alone on their line (see https://leanprover-community.github.io/contribute/style.html for style examples).

sgouezel commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,185 @@
+/-!
+# Isometries of the Complex Plane
+
+Classification of isometries in the complex plane.
```
Can you mention here the name of the Lean theorem below containing the classification, and also say what the classification is? I think it would also be worth explaining the proof strategy you will be following (with references to the Lean statements when it helps), to help the reader follow what is going on in the file.

sgouezel reviewed (commented):
Thanks for this! I have a lot of stylistic comments, but this is really the normal process :-)

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: code review fixes

frankymacster committed:
Merge branch 'linear_isometry_complex' of https://github.com/leanprover-community/mathlib into linear_isometry_complex

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: move out `id_apply` and `id_to_linear_map`

frankymacster committed:
Merge branch 'master' into linear_isometry_complex

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: simplify `simp`

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: remove unneeded comments

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: move `noncomputable theory` after docstring

sgouezel commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,183 @@
+      simp only [a],
+      rw linear_isometry.norm_map,
+      simp,
+    },
```
```suggestion
  { change ∥a∥ = 1,
    simp only [a],
    rw linear_isometry.norm_map,
    simp },
```
Braces should be aligned with the previous indentation, and the last brace should not be alone on its line. Same thing everywhere in this proof.

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: make last brace not alone on its line

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: align braces with previous indentation

fpvandoorn commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,171 @@
+  (f z).im = z.im ∨ (f z).im = -z.im :=
+begin
+  specialize h₁ z,
+  simp [complex.abs] at h₁,
```
Can you remove this [non-terminal simp](https://leanprover-community.github.io/extras/simp.html#non-terminal-codesimpcodes-and-codesimpacode)?
```suggestion
  simp only [complex.abs] at h₁,
```

francoissunatori replied:
thanks for the link with explanation :)

fpvandoorn commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,171 @@
+      norm_map' := by {
+        intros x,
+        simp,
+        change ∥f 1∥⁻¹ * ∥f x∥ = ∥x∥,
```
nonterminal simp
```suggestion
        suffices : ∥f 1∥⁻¹ * ∥f x∥ = ∥x∥, { simpa }
```

fpvandoorn commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,171 @@
+        change ∥f 1∥⁻¹ * ∥f x∥ = ∥x∥,
+        iterate 2 { rw linear_isometry.norm_map },
+        simp } },
+    have hg0 : g 0 = 0 := g.map_zero,
```
unused
```suggestion
```

francoissunatori replied:
oh thanks, I hadn't noticed that!

fpvandoorn commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,171 @@
+      change 0 < ∥a∥ ^ 2,
+      rw linear_isometry.norm_map,
+      simp,
+      apply zero_lt_one },
```
non-terminal simp
```suggestion
      simp [zero_lt_one] },
```

fpvandoorn commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,171 @@
+      change 0 < ∥a∥ ^ 2,
+      rw linear_isometry.norm_map,
+      simp,
+      apply zero_lt_one },
```
non-terminal simp
```suggestion
      simp [zero_lt_one] },
```

fpvandoorn commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,171 @@
+  { change ∥a∥ = 1,
+    simp only [a],
+    rw linear_isometry.norm_map,
+    simp },
```
```suggestion
  { simp only [← norm_eq_abs, a, linear_isometry.norm_map, norm_one] },
```

fpvandoorn commented on src/analysis/complex/isometry.lean:
```diff
@@ -0,0 +1,171 @@
+      skip,
+      find (_ = _) {
+        rw ← mul_right_inj' ha_inv, rw ← mul_assoc, rw inv_mul_cancel ha, rw one_mul } },
+    exact h },
```
After #7705 is merged this proof can be reduced to
```suggestion
    simpa only [← inv_mul_eq_iff_eq_mul' ha] }
```

fpvandoorn reviewed (commented):
Sorry that this review is taking so long!
I have some suggestions, one of them depends on #7705.
Other than that, this LGTM.

Commit:
Merge remote-tracking branch 'origin/master' into linear_isometry_complex

frankymacster committed:
Merge branch 'master' into linear_isometry_complex

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: code review fixes

frankymacster committed:
feat(analysis/complex/isometry):
linear_isometry_complex: fix broken `is_basis.ext`

frankymacster committed:
Merge branch 'linear_isometry_complex' of https://github.com/leanprover-community/mathlib into linear_isometry_complex
